src/TopOptEMFocus.jl

```diff
@@ -0,0 +1,550 @@
+# 
+# Note that here we are defining a "callable object" of type `Λ` that encapsulates all of the PML parameters. This is convenient, both because we can pass lots of parameters around easily and also because we can define additional methods on `Λ`, e.g. to express the `∇(Λv)` operation.
+# 
+
+R = 1e-10
+LHp=(L/2, h1+h2)   # Start of PML for x,y > 0
+LHn=(L/2, 0)       # Start of PML for x,y < 0
+phys_params = (; k, n_metal, n_air, μ, R, dpml, LHp, LHn)
+
+# PML coordinate streching functions
+function s_PML(x; phys_params)
+    σ = -3 / 4 * log(phys_params.R) / phys_params.dpml / phys_params.n_air
+    xf = Tuple(x)
+    u = @. ifelse(xf > 0 , xf - phys_params.LHp, - xf - phys_params.LHn)
+    return @. ifelse(u > 0,  1 + (1im * σ / phys_params.k) * (u / phys_params.dpml)^2, $(1.0+0im))
+end
+
+function ds_PML(x; phys_params)
+    σ = -3 / 4 * log(phys_params.R) / phys_params.dpml / phys_params.n_air
+    xf = Tuple(x)
+    u = @. ifelse(xf > 0 , xf - phys_params.LHp, - xf - phys_params.LHn)
+    ds = @. ifelse(u > 0, (2im * σ / phys_params.k) * (1 / phys_params.dpml)^2 * u, $(0.0+0im))
+    return ds.*sign.(xf)
+end
+
+struct Λ{PT} <: Function
+    phys_params::PT
+end
+
+function (Λf::Λ)(x)
+    s_x,s_y = s_PML(x; Λf.phys_params)
+    return VectorValue(1/s_x, 1/s_y)
+end
+
+# Define the derivative for the Λ factor
+Fields.∇(Λf::Λ) = x -> TensorValue{2, 2, ComplexF64}(-(Λf(x)[1])^2 * ds_PML(x; Λf.phys_params)[1], 0, 0, -(Λf(x)[2])^2 * ds_PML(x; Λf.phys_params)[2])
+
+
+# ## Filter and threshold
+# 
+# Here we use the filter and threshold dicussed above. The parameters for the filter and threshold are extracted from Ref [6]. Note that every integral in the filter is only defined on $\Omega_d$
+# 
+
+r = 5/sqrt(3)               # Filter radius
+β = 32.0                    # β∈[1,∞], threshold sharpness
+η = 0.5                     # η∈[0,1], threshold center    
+
+a_f(r, u, v) = r^2 * (∇(v) ⋅ ∇(u))
+
+function Filter(p0; r, fem_params)
+    ph = FEFunction(fem_params.P, p0)
+    op = AffineFEOperator(fem_params.Pf, fem_params.Qf) do u, v
+        ∫(a_f(r, u, v))fem_params.dΩ_d + ∫(v * u)fem_params.dΩ_d, ∫(v * ph)fem_params.dΩ_d
+      end
+    pfh = solve(op)
+    return get_free_dof_values(pfh)
+end
+
+function Threshold(pfh; β, η)
+    return ((tanh(β * η) + tanh(β * (pfh - η))) / (tanh(β * η) + tanh(β * (1.0 - η))))         
+end
+
+
+# ## Weak form
+# 
+# We notice that the design parameters only affect the weak form in the design domain and the PML does not affect the design domain, we can then make things simpler by dividing the weak form to a base integral that contains the whole computation cell and an additional integral on the design domain. We also make a LU factorization on the final Maxwell operator matrix $A$ since it will only be used to solve for linear equations. 
```

Comment:
```suggestion
# Notice that the design parameters only affect the weak form in the design domain, whereas the PML does not affect the design domain. Therefore, we can then make things simpler by dividing the weak form into a base integral that applies to the whole computation cell and an additional integral on the design domain that only includes the $\varepsilon \ne 1$ terms. We return only the LU factorization of the final Maxwell operator matrix $A$, since that's all we need to solve the resulting equatiohns.
```